main/lsp/protocol.cc

```diff
@@ -277,9 +281,21 @@ optional<unique_ptr<core::GlobalState>> LSPLoop::runLSP() {
                     }
                 }
                 msg = move(processingQueue.pendingRequests.front());
-                exitProcessed = msg->isNotification() && msg->method() == LSPMethod::Exit;
                 processingQueue.pendingRequests.pop_front();
                 hasMoreMessages = !processingQueue.pendingRequests.empty();
+
+                if (msg->isNotification()) {
+                    const auto method = msg->method();
+                    exitProcessed = method == LSPMethod::Exit;
+                    if (opts.lspCancelableSlowPathEnabled && method == LSPMethod::SorbetWorkspaceEdit) {
+                        // While we're holding the queue lock (and preventing new messages from entering), start a
+                        // commit for an epoch if this message will trigger a cancelable slow path.
+                        const auto &params = get<unique_ptr<SorbetWorkspaceEditParams>>(msg->asNotification().params);
+                        if (!params->updates.canTakeFastPath) {
+                            gs->startCommitEpoch(params->updates.versionEnd);
```

Comment:
This is the only slightly lurky bit. There are ENFORCEs to ensure that we call this method before trying to commit an edit with cancelation enabled.

Reply:
why does this have to be guarded with `opts.lspCancelableSlowPathEnabled`?
Can we keep the logic running without it?

Reply:
Good question. I guess we don't have to, and we can just enforce that cancelation doesn't happen in the preprocessor!